[2단계 - 출석 다시 구현하기] 부기(이창의) 미션 제출합니다.

changuii opened the pull request:
<!-- 

## 코드 리뷰 팁

- 코드와 관련된 질문이 있다면, PR 본문에 적기 보다는 해당 코드를 선택하고 코멘트를 남겨주세요.
  - [참고: Adding comments to a pull request](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/commenting-on-a-pull-request#adding-comments-to-a-pull-request)

-->

## 체크 리스트

- [x] 미션의 필수 요구사항을 모두 구현했나요?
- [x] Gradle `test`를 실행했을 때, 모든 테스트가 정상적으로 통과했나요?
- [x] 애플리케이션이 정상적으로 실행되나요?
- [x] [프롤로그](https://prolog.techcourse.co.kr)에 셀프 체크를 작성했나요?
  - [프롤로그 셀프체크 링크](https://prolog.techcourse.co.kr/studylogs/3953)


## 객체지향 생활체조 요구사항을 얼마나 잘 충족했다고 생각하시나요?

### 1~5점 중에서 선택해주세요.

- [ ] 1 (전혀 충족하지 못함)
- [ ] 2
- [ ] 3 (보통)
- [x] 4
- [ ] 5 (완벽하게 충족)

## 선택한 점수의 이유를 적어주세요.
### 규칙 1: 한 메서드에 오직 한 단계의 들여쓰기(indent)만 한다.
- try catch 이외의 부분에서는 한 단계의 indent로만 작성했습니다.
### 규칙 2: else 예약어를 쓰지 않는다.
- else 예약어를 사용하지 않았습니다.
### 규칙 3: 모든 원시값과 문자열을 포장한다.
- 의미가 있는 원시값과 문자열을 포장하였습니다. crewName은 현재 굳이 할 필요 없다고 생각해 포장하지 않았습니다.
### 규칙 4: 한 줄에 점을 하나만 찍는다.
- 한 줄에 점을 하나만 찍었습니다.
- 테스트 환경에서는 여러 점을 사용했습니다.
### 규칙 5: 줄여쓰지 않는다(축약 금지).
- 변수와 메서드 명을 최대한 풀어 작성했습니다.
### 규칙 6: 모든 엔티티를 작게 유지한다.
- TDD로 구현하면서 엔티티들이 작게 분리되었습니다.
### 규칙 7: 3개 이상의 인스턴스 변수를 가진 클래스를 쓰지 않는다.
- 3개 이상의 인스턴스 변수를 가진 클래스가 없습니다.
### 규칙 8: 일급 콜렉션을 쓴다.
-crews, AttendanceBook, AttendanceHistory가 일급 콜렉션입니다.
### 규칙 9: 게터/세터/프로퍼티를 쓰지 않는다.
- 세터는 사용하지 않았습니다.
- 게터는 테스트 용도 혹은 뷰에게 데이터를 전달하는 용도로만 사용하였습니다.
- 외부에서 값을 꺼내 로직을 수행하는 부분이 없습니다.

  
## 어떤 부분에 집중하여 리뷰해야 할까요?

TDD 사이클에 대한 리뷰를 받고 싶어 red -> green -> refactor 과정으로 커밋했습니다. 이 부분을 집중해서 리뷰 부탁드립니다.

### 궁금한 점
TDD를 하면서 큰 기능을 테스트로 작성하려다 보니 많은 세부 클래스들이 필요했습니다.
저는 큰 기능을 테스트하려다가 세부 기능들이 많이 필요한 경우 기능 정의를 더 세분화해서 작성한 후, 작은 단위의 테스트를 진행하고, 큰 기능으로 다시 돌아왔습니다.
하지만, 이렇게 구현하다 보니 작은 기능을 구현하고 돌아왔을때, 어느 부분을 해야하는지 까먹는 포인트가 많았습니다.
로키는 이런 경우 어떻게 하시는지 궁금합니다~~! 🧐

changuii committed:
init: project init

changuii committed:
docs: docs 초기화

changuii committed:
docs(README) : 기능 요구사항을 정리하라

changuii committed:
red : 닉네임에 해당하는 크루를 반환한다.

- 레드 테스트를 작성하였다.

changuii committed:
green: 닉네임에 해당하는 크루를 가져오라

- 닉네임에 해당하는 크루를 반환하도록 하였다.

changuii committed:
red: 닉네임에 해당하는 크루가 존재하지 않는다면, 예외를 발생시켜라

changuii committed:
green: 닉네임에 해당하는 크루가 존재하지 않는다면, 예외가 발생시켜라

changuii committed:
docs(README): 구현한 기능에 대해 체크하라

changuii committed:
red: 새로운 출석 기록(오늘 날짜에, 입력받은 등교시간)을 추가하라

changuii committed:
green: 새로운 출석 기록(오늘 날짜에, 입력받은 등교시간)을 추가하라

changuii committed:
red: 등교 날짜가 아닌 경우 예외가 발생한다.

changuii committed:
green: 등교 날짜가 아닌 경우 예외가 발생한다.

changuii committed:
refactor: 등교 날짜가 아닌 경우 예외가 발생한다.

changuii committed:
red: 이미 출석한 경우라면, 예외가 발생한다.

- IllegalStateException으로 잘못된 호출임을 알려준다.

changuii committed:
green: 이미 출석한 경우라면, 예외를 발생시켜라

changuii committed:
refactor: 이미 출석한 경우라면, 예외가 발생한다.

changuii committed:
red: 크루에 해당하는 출석 기록을 가져온다.

changuii committed:
green: 크루에 해당하는 출석 기록을 가져온다.

changuii committed:
red: 이미 존재하는 크루에 대해서 새롭게 register하면, 예외가 발생시켜라

changuii committed:
green: 이미 존재하는 크루에 대해서 새롭게 register하면, 예외가 발생시켜라

changuii committed:
refactor: 이미 존재하는 크루에 대해서 새롭게 register하면, 예외가 발생시켜라

changuii committed:
red: 존재하지 않는 크루에 대해 AttendanceHistory를 요구하면, 예외가 발생시켜라

changuii committed:
green: 존재하지 않는 크루에 대해 AttendanceHistory를 요구하면, 예외가 발생시켜라

changuii committed:
refactor: 존재하지 않는 크루에 대해 AttendanceHistory를 요구하면, 예외가 발생시켜라

changuii committed:
docs(README): 구현된 사항을 체크하라

changuii committed:
docs(README): 출석 수정 기능을 좀 더 자세하게 세분화하라

changuii committed:
red: 기존의 출석 기록을 새로운 시간으로 수정한다.

changuii committed:
docs(README): 출석 업데이트 로직을 더 세분화하라

changuii committed:
red: 변경된 출석 기록을 가져온다.

changuii committed:
red: 변경된 출석 기록을 가져온다.

- 매개변수를 추가하였다.

changuii committed:
green: 변경된 출석 기록을 가져온다.

changuii committed:
green: 기존의 출석 기록을 새로운 시간으로 수정하고 이전 값을 반환한다.

changuii committed:
docs(README): 기록이 존재하지 않는 경우에 대한 로직 추가

changuii committed:
red: 등교 날짜가 아닌 날짜의 출석 기록을 찾는다면, 예외를 발생시켜라.

changuii committed:
green: 등교 날짜가 아닌 날짜의 출석 기록을 찾는다면, 예외가 발생시켜라

changuii committed:
red: 등교 날짜가 아닌 날짜의 출석을 수정한다면, 예외를 발생시켜라

changuii committed:
green: 등교 날짜가 아닌 날짜의 출석을 수정한다면, 예외를 발생시켜라

changuii committed:
red: 수정하려는 시간이 등교 시간이 아니라면, 예외를 발생시켜라

changuii committed:
green: 수정하려는 시간이 등교 시간이 아니라면, 예외를 발생시켜라

changuii committed:
refactor: 수정하려는 시간이 등교 시간이 아니라면, 예외를 발생시켜라

changuii committed:
red: 기록이 존재하지 않는, 날짜의 경우에는 empty record를 반환하라

changuii committed:
green: 기록이 존재하지 않는, 날짜의 경우에는 empty record를 반환하라

changuii committed:
docs(README): 구현한 기능을 체크하라

changuii committed:
docs(README) 크루별 출석 기록 확인을 더 세세하게 작성하라

changuii committed:
red: 전날까지의 출석 기록을 empty를 포함해서 반환하라

changuii committed:
green: 전날까지의 출석 기록을 empty를 포함해서 반환한다.

changuii committed:
refactor: 전날까지의 출석 기록을 empty를 포함해서 반환하라

changuii committed:
docs(README): 출석 통계 계산을 더 세분화하여 작성한다.

changuii committed:
red: 출석의 상태를 계산하라

changuii committed:
green: 출석의 상태를 계산하라

changuii committed:
refactor: 출석의 상태를 계산하라

changuii committed:
red(AttendanceRecord): 출석의 상태를 계산한다.

changuii committed:
chore: 테스트들의 패키지를 도메인 패키지로 옮겨라

changuii committed:
green(AttendanceRecord): 출석의 상태를 계산한다.

changuii committed:
red(AttendanceHistory): 전날까지의 출석 통계를 계산하라

changuii committed:
red: record가 비어있다면, ABSENCE를 반환한다.

changuii committed:
green: record가 비어있다면, ABSENCE로 처리하라

changuii committed:
green: 전날까지의 출석 통계를 계산하여 반환하라

changuii committed:
red: 제적 위험자인지 계산하라

changuii committed:
fix: calcualteRiskOfExpulsion 인자로 크루를 받도록 하라

changuii committed:
red: 제적 위험자인지 판단하라

changuii committed:
green: 제적 위험자인지 판단하라

changuii committed:
refactor: 제적 위험자인지 판단하라

changuii committed:
test(RiskOfExpulsionStatus): 음수가 전달되면 예외가 발생한다.

changuii committed:
green: 제적 위험자인지 계산하라

changuii committed:
fix: 정렬 기준이 내림차순이지만, 오름 차순으로 메서드 이름이 명명되어 있는 것을 수정하라

changuii committed:
red: 제적 위험자들을 반환하라

changuii committed:
red(AttendanceHistory): 제적 위험이 존재하는지를 반환한다.

changuii committed:
green: 제적 위험이 존재하는지 반환하라

changuii committed:
red: 제적 위험 상태 반환의 책임을 History로 이동시켜라

changuii committed:
green: 제적 위험 상태를 반환하라

changuii committed:
refactor: 제적 위험 상태를 반환하라

changuii committed:
green : 제작 위험자들의 통계를 반환하라

changuii committed:
chore: 불필요한 import와 줄바꿈을 제거하라

changuii committed:
refactor: 명칭들 통일하고, getter()를 하단으로 이동시켜라

changuii committed:
refactor(AttendanceHistory): 요일의 검증을 유틸로 분리하였다.

- 어디서나 통용되는 부분에 대해서 유틸로 분리하였다.

changuii committed:
refactor(AttendanceHistory): 공휴일 검증의 역할을 SolarCalendarHoliday로 분리하였다.

changuii committed:
refactor: 불필요한 변수를 제거하라

changuii committed:
feat: 크루별 출석 기록 확인을 기능을 구현하라

changuii committed:
refactor(AttendanceHistory): isRiskOfExpulsion의 중복되는 연산을 제거하라

changuii committed:
red: statistics가 아닌, AttendanceHistory로 변경하라

changuii committed:
green: statistics가 아닌, AttendanceHistory로 변경하라

changuii committed:
red: calculateRiskOfExpulsionHistory 메서드로 변경하라

changuii committed:
green: calculateRiskOfExpulsionHistory 메서드로 변경하라

changuii committed:
feat: 제적 위험자 확인 로직을 작성하라

changuii committed:
feat: 출석 확인기능을 구현하라

changuii committed:
feat: 출석 수정을 구현하라

changuii committed:
docs(README): 완료된 사항을 체크하라

changuii committed:
test: 전체적으로 경계값에 대한 테스트를 추가하라

changuii committed:
feat: 사용자가 Q를 입력하기 전까지 기능을 수행하라

changuii committed:
feat: 예외 발생시 재 입력을 받도록 하라

changuii committed:
feat: 도메인 로직 중, 예외 발생시 메시지를 사용자에게 안내하고 다시 Command를 시작한다.

changuii committed:
refactor: 출력에 줄바꿈 추가

changuii committed:
docs(README): 완료된 기능을 체크하라

changuii committed:
feat: 이미 출석이 존재하는 경우, 수정 기능을 사용하도록 안내하라

changuii committed:
test(AttendanceHistoryTest): 주어진 날짜에 이미 출석이 존재하는지 여부를 반환하라

Rok93 commented on README.md:
````diff
@@ -2,27 +2,320 @@
+Q
+```
+
+### 프로그래밍 요구 사항
````
> (질문) TDD를 하면서 큰 기능을 테스트로 작성하려다 보니 많은 세부 클래스들이 필요했습니다.
저는 큰 기능을 테스트하려다가 세부 기능들이 많이 필요한 경우 기능 정의를 더 세분화해서 작성한 후, 작은 단위의 테스트를 진행하고, 큰 기능으로 다시 돌아왔습니다.
하지만, 이렇게 구현하다 보니 작은 기능을 구현하고 돌아왔을때, 어느 부분을 해야하는지 까먹는 포인트가 많았습니다.
로키는 이런 경우 어떻게 하시는지 궁금합니다~~! 🧐

정확한 상황은 제가 추측할 수 밖에 없긴하지만.... 

결국 저는 기능 구현 목록표를 작은 단위로 잘 작성하는게 일차적으로 중요하다고 생각해요. 

이렇게 기능 구현 목록표를 더 세부적으로 쪼갤수록 얻을 수 있는 장점도 생긴다고 생각해요. 
그 장점으로는 작은 단위일수록 테스트 케이스를 작성하기가 더 쉬워진다고 생각해요. (별개로 커밋단위도 적어지다보니 리뷰어가 리뷰하기에도 수월해지구요. (아마 그 외에도 작은 단위가 여러모로 부담감이 적기도하고, 절대 나쁠건 없다고 생각해요 😃)

그렇다면 처음부터 기능 구현 목록표를 잘 짜야하는가...? 라고한다면 저는 NO 라고 생각합니다. 
물론 처음부터 잘 쪼개시는 분도 있을 수 있지만, 대다수는 처음 정의했던 것과 달리 변경사항이 발생할 수 있다고 생각해요. 
이런 면에서 저는 기능 구현 목록표는 유동적이라 생각하고, 기능을 구현하면서 더 세부적인 단위로 쪼개지면서 커밋 단위도 더 쪼개질 수 있다고 생각해요.  

필요하다면 더 세부적인 단위로 기능을 나누고 지금 작성하는 코드는 원래 작서하려던 큰 단위가 아닌 작은 단위로 넘어가서 TDD 사이클을 돌아야한다고 생각합니다. 😃 

> 하지만, 이렇게 구현하다 보니 작은 기능을 구현하고 돌아왔을때, 어느 부분을 해야하는지 까먹는 포인트가 많았습니다.
로키는 이런 경우 어떻게 하시는지 궁금합니다~~! 

저는 주로 `TODO` 주석을 많이 활용하는 편입니다. 
(플러그인 영향인지 정확히는 모르겠지만...) 커밋을 할 때, TODO 주석이 남아있는 경우 경고를 해주기 때문에 이부분을 꼭 체크하는 습관이있습니다. 이렇게 확인하다보면 우려하시는 상황은 잘 발생하지 않더라구요. 😃

그 외에도 TDD는 결국 자신만의 사이클을 만들어가는게 중요한 것 같아요. 
저도 이방식 저 방식 조언을 구해서 들어본 방식들은 여러개있지만, 저한테 맞는 방식은 
결국 이런저런 방식들을 잘 조합하여 만든 별개의 방식이었던 것 같아요. (그 마저도 완벽하다고 생각하지는 않습니다. 😅)
앞으로 부기가 이런 불편함을 느꼈으니, 이 부분을 늘 염두해보면서 어떤 포인트가 더 좋은지, 저 말고도 다른 리뷰어, 혹은 페어에게 질문해가면서 자신에게 맞는 방식을 찾아가는 것이 더더욱 도움이 많이 되실거라 생각해요. 😃

(너무 좋은 질문을 해주셔서 저도 모르게 장문의 코멘트를 남겼네요. 😅)

Rok93 commented on src/main/java/components/DtoConverter.java:
```diff
@@ -0,0 +1,79 @@
+package components;
```
`components` 라는 패키지는 어떤의도로 네이밍하신걸까요? 🤔

changuii replied:
도메인 패키지에 속하지 않는 클래스들을 담아두기 위해 만들었습니다!

해당 패키지의 클래스들은 도메인으로써 여러 도메인 객체들과 상호작용하지 않고, '저장된 데이터를 읽어오는 객체' 혹은 '도메인 객체를 DTO 객체로 변환'과 같은 기능만 수행합니다.
이는, 책임을 가지고 어떤 행위를 통해 중요한 비즈니스 로직들을 수행한다기 보다는, 애플리케이션의 흐름을 위해 필요하다고 생각했습니다.!

로키의 생각이 궁금합니다. 
- 이와 같은 경우에는 어떤 패키지 명이 좋을까요? 
- 애플리케이션의 기능 ex) 크루의 출석 확인, 크루의 출석 수정, 크루의 출석 내역 확인... 등과 같은 기능이 비즈니스 로직이라고 생각하시나요 아니면, 전체 애플리케이션의 모든 로직이 비즈니스 로직이라고 생각하시나요?

Rok93 replied:
> 도메인 패키지에 속하지 않는 클래스들을 담아두기 위해 만들었습니다!

도메인 패키지에 속하지 않는 경우에는 모두 `components`에 두는걸까요? 
개인적으로는 그런 경우라면 external이 더 올바른 표현이 아닐까 싶어요. 🤔

추후 정말로 도메인 외부의 케이스(ex. 외부 결제 API, 카카오, 네이버 등의 API etc...)가 추가된다면 이 패키지가 조금 애매해지지 않을까 싶은데요. 🤔 

DtoConverter는 `모든 도메인 객체 --> DTO`하는 로직을 관리하는 곳으로 보이는데, 
추후 DTO가 늘어나면 늘어날 수록 이 DtoConverter의 크기는 방대해질 것 같네요. 👀  
이렇게되면 추후 팀원들이 각자 자신의 작업을 수행하게될테고 이떄 DtoConverter를 여러 작업자들이 동시에 수정하기 때문에 깃 충돌도 많이 발생할 것 같아요. 🤔 

Converter를 별도로 두기보다는,  DTO 객체의 정적 패토리 메서드로 도메인 객체를 받아 DTO로 변환하도록 해보면 어떨까요? 

또한 `SavedDataLoader`는 결국 `Crews`의 (Crew 각각의 AttendancyBook 포함)초기 데이터를 셋팅해주는 역할이라 생각되는데, 
조금 더 나아가서 SavedDataLoader를 `CrewsFactory`로 명명해서 Crews 객체를 생성하는 역할을 위임해볼 수도 있을 것 같아요!  

(이렇게 한다면....!?  components 패키지가 없어질 것 같네요 🙄)

> 애플리케이션의 기능 ex) 크루의 출석 확인, 크루의 출석 수정, 크루의 출석 내역 확인... 등과 같은 기능이 비즈니스 로직이라고 생각하시나요 아니면, 전체 애플리케이션의 모든 로직이 비즈니스 로직이라고 생각하시나요?

이 부분은 전체 애플리케이션이 비즈니스 로직이라고 생각하지는 않고 말씀하신 내용중 선택하라면 전자에 가까울 것 같아요.  🤔

changuii replied:
dto의 경우에 Data Transfer Object가 기능을 가지면 안되겠다고 판단하여 그렇게 작성했는데, 말씀하신 대로 Converter의 책임이 방대해지면 방안을 세울 수 없다고 판단하여 Dto가 변환 작업을 가지도록 변경하였습니다,.
- DtoConverter에 있던 기본타입, 컬렉션 변환 작업은 controller로 이동시켰습니다.

CrewsFactory보다는, AttendanceSavedDataLoader가 더 직관적인 이름이라고 판단하여 이름을 그렇게 변경하였고, 해당 객체가 도메인 로직을 수행한다고 판단하여 패키지를 옮겨주었습니다.

Rok93 commented on src/main/java/components/SavedDataLoader.java:
```diff
@@ -0,0 +1,62 @@
+    private static final int ATTENDANCE = 1;
+    private static final String DELIMITER = ",";
+    private static final DateTimeFormatter FORMATTER = DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm");
+    private final List<String[]> loadedData;
```
사소한 부분일 수 있겠지만, 상수와 상태는 줄바꿈을 통해서 가독성을 높여주면 좋을 것 같아요.
```suggestion
    private static final DateTimeFormatter FORMATTER = DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm");
    
    private final List<String[]> loadedData;
```

Rok93 commented on src/main/java/components/SavedDataLoader.java:
```diff
@@ -0,0 +1,62 @@
+    private static final int CREW = 0;
+    private static final int ATTENDANCE = 1;
+    private static final String DELIMITER = ",";
+    private static final DateTimeFormatter FORMATTER = DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm");
```
FORMATTER, DELIMITER 와 같이 포괄적인 네이밍보다는, 조금 더 의도를 표현할 수 있는 네이밍을 사용하면 좋을 것 같아요. 🤔
비슷한 FORMATTER와 DELIMITER가 여러개인 경우에는 해당 네이밍으로 구분이 가능할지 고민해보면 될 것 같아요. 😃

Rok93 commented on src/main/java/components/SavedDataLoader.java:
```diff
@@ -0,0 +1,62 @@
+        attendanceHistory.attendance(parseDateTime(line[ATTENDANCE]));
+    }
+
+    private void registerCrew(final AttendanceBook attendanceBook, final Crew crew) {
```
음.. 개인적인 생각에서는 `registerCrew`,  `loadAttendances` 행위를 별도의 행위로 쪼갤 필요가 있을까 생각이 들어요 
SavedDataLoader라면 데이터를 로드해서 읽는 역할을 가졌다고 생각하는데, 사실 초기 데이터를 읽어서 저장한다와 같이 한번에 두 행위를 같이 수행해도 충분할 것 같아서요. 🤔

changuii replied:
다시 코드를 보니, 확실히 조금 부자연스러운 것 같아요
- 현재, attendances.csv 밖에 없는데, 제가 설계한 구조는 crews와 attendanceBook이 존재하는데 이것 때문에 간극이 생긴 것 같아요

그래서 생각한 방법은, AttendanceBook을 만들고난 후, crews를 통해 초기화하는 과정이 있으면 좋을 것 같다고 생각했어요 ~~
- 크루들의 목록을 가지고, 출석부를 초기화!

Rok93 commented on src/test/java/domain/CrewsTest.java:
```diff
@@ -0,0 +1,50 @@
+public class CrewsTest {
+
+    @Nested
+    @DisplayName("성공 테스트")
```
[BDD](https://mingule.tistory.com/43) 스타일로 테스트 코드를 작성해보면 어떨까 싶어요. 🤔

더 나아가서  지금처럼 성공, 실패와 같은 이분법적인 방식보다는, 

행위에 따른 계층적 테스트 코드를 작성하는 것이 조금 더 테스트 코드의 유지보수성에 도움이 되지 않을까 싶어요, 🤔

Rok93 commented on src/main/java/domain/AttendanceBook.java:
```diff
@@ -0,0 +1,50 @@
+                .stream()
+                .map(Entry::getValue)
+                .filter(attendanceHistory -> attendanceHistory.isRiskOfExpulsion(targetDate))
+                .collect(Collectors.toList());
```
아래와 같이 변경해볼 수 있을 것 같네요. 😉 
```suggestion
        return attendanceBook.values()
                .stream()
                .filter(attendanceHistory -> attendanceHistory.isRiskOfExpulsion(targetDate))
                .collect(Collectors.toList());
```

Rok93 commented on src/main/java/domain/AttendanceHistory.java:
```diff
@@ -0,0 +1,121 @@
+    }
+
+    public AttendanceRecord updateTimeByDate(final LocalDateTime afterTime) {
+        validateAttendanceDay(afterTime);
+        validateAttendanceTime(afterTime);
```
이 둘을 매번 출석과 관련된 유효성 검증에서는 셋트로 묶여서 수행되는 것 같은데, 이런 케이스는 하나로 묶어도 되지 않을까 싶네요.😃

changuii replied:
```java
validateAttendanceDay(afterTime);
validateAttendanceTime(afterTime);
```

이 둘을 동시에, 검증하는 곳이 updateTimeByDate() 밖에 없는데 굳이 묶어서 처리하는 것보다는 직관적으로 분리해두는게 더 가독성이 좋다고 생각합니다.
이렇게 두고 나중에 여러 곳에서 재사용하는 경우가 되면 묶어서 작성하는게 더 낫다고 생각하는데, 로키의 생각이 궁금합니다! ~ 🧐

Rok93 replied:
개인적인 생각으로는 저는 출석을 신규로하거나 혹은 업데이트할 때,  즉 출석 행위에 대한 검증이라 생각이되어서 `validateAttendance` 라는 메서드를 둘 것 같아요.  (현재 attendance와 updateTimeByDate 두 군데서 동일하게 쓰이네요)

지금처럼 분리해두게되면, 누군가는 출석에 대한 신규 행위가 추가되었을 때, validateAttendanceDay, validateAttendanceTime 둘 중 하나를 누락한다거나하는 실수를 할 여지가 생긴다고 생각해요. 🤔

changuii replied:
맞아요 첫 피드백에는 한 곳에만 있었는데 말씀해주신 것처럼 테스트를 행위 기준으로 정리하다 보니, 빠진 것을 발견해서 수정했어요!
반영하겠습니다.

Rok93 commented on src/main/java/domain/AttendanceHistory.java:
```diff
@@ -0,0 +1,121 @@
+        validateAttendanceDay(attendanceDateTime);
+        validateAlreadyAttendance(attendanceDateTime);
+        final AttendanceRecord attendanceRecord = new AttendanceRecord(attendanceDateTime);
+        attendanceHistory.put(attendanceDateTime.toLocalDate(), attendanceRecord);
```
`updateTimeByDate`의 경우에는 이미 출석을 했더라도 수정하는 것이니 덮어쓰기를하면 될 문제지만, (신규로) 출석하는 것은 이미 출석한 케이스에 대한 적절한 예외처리가 필요하지 않을까 생각이드네요.  (당연히 이 케이스가 추가된다면 예외 케이스에 대한 테스트 코드도 추가가 필요할 것 같구요! 👀)

changuii replied:
`validateAlreadyAtttendance`를 통해, 이미 출석한 케이스에 대해서 검증을 거쳤습니다!

Rok93 commented on src/main/java/domain/AttendanceHistory.java:
```diff
@@ -0,0 +1,121 @@
+                || SolarCalendarHoliday.isHoliday(MonthDay.from(targetDate))) {
+            return false;
+        }
+        return true;
```
boolean을 반환하는 타입은 아래와 같이 if 문 없이도 해결되는 경우가 꽤 많답니다. 😃 

```suggestion
        return !DayOfWeeks.isWeekend(targetDate.getDayOfWeek())
            && !SolarCalendarHoliday.isHoliday(MonthDay.from(targetDate));
```

Rok93 commented on src/main/java/domain/AttendanceRecord.java:
```diff
@@ -0,0 +1,41 @@
+    }
+
+    public static AttendanceRecord empty(final LocalDate date) {
+        final LocalDateTime emptyTime = LocalDateTime.of(date, LocalTime.NOON);
```
출석 기록이 없는 경우 날짜에 대해서는 `12:00`에 출석을 기록하는게 맞을까요? 
물론 `isEmpty`가 있긴하지만, 임의의 `attendanceDateTime`를 두시는 것에 대해서는 조금 우려되는 점이있는 것 같아요. 🤔
ex. 월요일과 같이 출석 시간이 13:00 인 경우에는 12:00의 출석 시간은 정상 출석으로 인정?

Rok93 commented on README.md:
```diff
@@ -2,27 +2,320 @@
 
 출석 미션 저장소
 
-### 1. 파일 입출력
```
> TDD 사이클에 대한 리뷰를 받고 싶어 red -> green -> refactor 과정으로 커밋했습니다. 이 부분을 집중해서 리뷰 부탁드립니다.

엄밀하게 TDD 사이클을 지켰는지는 커밋 내용을 통해서만 확인이 가능할 것 같은데요. 😃
우려하신 것과는 달리 TDD(red -> green -> blue) 싸이클을 잘 지켜주신걸로 보이네요. 😉 

잘하고 계시다고 생각해요.

Rok93 reviewed (changes requested):
부기 2단계도 빠르게 잘 구현해주셨습니다. 👍👍 
몇몇 코멘트 남겨두었는데 확인 부탁드리게요. 🙏

changuii committed:
refactor(SavedDataLoader): 변수명을 축약하지 않고 자세하게 작성하라

changuii committed:
refactor: 의미있는 줄바꿈을 통해 가독성을 높여라

changuii committed:
refactor(AttendanceBook): 의미없는 map()을 제거하라

- entrySet()을 stream하지말고, values()를 stream하라

changuii committed:
refactor(AttendanceHistory): 불필요한 if문을 제거하라

- boolean을 반환하는 경우, if로 처리하는 것보다는 값을 그대로 반환하는게 더 직관적이다.

changuii committed:
refactor: attendanceBook에 크루들을 등록하는 책임을 Crews 객체로 옮겨라

- SavedDataLoader가 데이터에서, 크루들을 등록한다기 보다는, "출석부"가 "크루의 목록"에게 크루를 등록해달라고 요청하는 것이 더 자연스럽다고 생각했다.

changuii committed:
refactor: 비어있는 AttendanceRecord의 경우, attendanceTime이 12시로 설정되어 있던 부자연스러움을 해결하라

- isEmpty()를 삭제했다.
- LocalDateTime을 분리하여 관리하도록 했다.
- attendanceTime에 대해 null(값이 진짜 없음)을 허용하되, 외부에서 안전하게 사용하도록 Optional로 감싸 반환하도록 변경하였다.

changuii committed:
test: 단정문이 많이 존재한다면, assertSoftly로 감싸 처리하도록 하라

changuii committed:
refactor(AttendanceBookTest): 성공, 실패 케이스로 테스트를 나누지 말고, 행위에 따라 나누어라

- 성공, 실패 테스트로 케이스를 나누게 되면, 코드가 정리되지 않아 어떤 부분에 대해서 테스트를 하지 않았는지 쉽게 발견하기 어려웠다.
- 하지만, 행위를 기준으로 테스트를 나누니 더 명확하게 무엇이 부족한지 한눈에 들어왔다.
- 객체의 구성이 행위를 중심으로 나타내기 때문에 이것이 더 명확하다고 생각한다.
- 만약 많은 테스트케이스가 생긴다면, 내부에서 한 번더 계층으로 나누면 될 것이다.

changuii committed:
refactor(AttendanceRecordTest): 행위를 기준으로, 테스트들을 계층화하라

changuii committed:
refactor(AttendanceStatusTest): 행위를 기준으로 계층화하여 작성하라

changuii committed:
refactor(CrewsTest): 행위를 기준으로 계층화하여 작성하라

changuii committed:
refactor(RiskOfExpulsionStatusTest): 행위를 기준으로 계층화하여 작성하라

changuii committed:
test(AttendanceHistoryTest): 행위를 기준으로 계층화하여 작성하라

changuii committed:
test(AttendanceHistory): 출석 시간이 아닌 경우, 예외가 발생하도록 하라

- 빠뜨린 테스트를 추가하라

changuii committed:
test(AttendanceHistoryTest): 기록이 없는 날짜의 출석을 찾는다면, 올바르게 Time이 빈 Record를 반환하라

- 빠뜨린 테스트를 추가했다.

changuii reviewed (commented):
리뷰 감사합니다. 😊
의문이 든 부분을 제외하고는 리뷰를 반영하였습니다.

### 변경점
1. SaveDataLoader의 변수명들을 축약하지 않고 풀어서 작성한다.
2. 줄바꿈을 의미있게 사용하여 가독성을 향상시켰다.
3. AttendanceBook에서, entrySet()으로 stream을 하여 발생한 의미없는 map() 동작을 제거하였다.
4. AttendanceHistory에서, boolean 타입을 반환할때 의미없는 if()문을 제거하고 그대로 반환하도록 변경하였다.
5. AttendanceBook의 crew 등록의 책임을 SavedDataLoader에서, Crews로 이동시켰다.
6. 비어있는 AttendanceRecord가 12시라는 시간을 가진 것의 부자연스러움을 해소하였다.
7. 많은 단정문이 존재하는 테스트가 실패했을 경우 실패 지점을 빠르게 찾기 위해 assertSoftly로 묶어서 처리하도록 수정하였다.
8. 테스트를 "성공" 혹은 "실패"로 나누어서 작성하지 않고 객체의 "행위"로 묶어 가독성을 향상시켰다.
9. 빠뜨린 테스트 케이스들을 추가하였다.
10. AttendanceRecord가 불변 객체이기 때문에 캐싱처리를 하여 메모리 효율을 높였다.

changuii committed:
docs(TEST_SCENARIO): 사용자 기준에서의 애플리케이션 사용 시나리오를 작성하라

changuii commented:
추가적으로, BDD 관점에서 TEST_SCENARIO를 작성해보았습니다. 이 부분에 대해서도 리뷰 부탁드리겠습니다.

BDD 시나리오에 따라 테스트를 진행하려고 했으나, 미션의 요구사항 중에 UI 로직은 테스트에서 제외하기도 했고, 테스트 시나리오에 따라 통합 테스트를 작성하는 것이 시간이 많이 들어 이 부분은 빼고 리뷰 요청 드립니다!!
- 제가 시간이 부족한 탓도 있어서 다른 부분(자바, 객체지향, TDD)에 집중해서 학습하고 이 부분은 다음 미션에서 설계할때 적용해서 학습해보겠습니다. 감사합니다 :)

changuii committed:
refactor(AttendanceRecord): 불변 객체에 대해서, 캐싱 처리를 하였다.

Rok93 commented on src/main/java/components/DtoConverter.java:
```diff
@@ -0,0 +1,79 @@
+        if (riskOfExpulsionStatus == RiskOfExpulsionStatus.WARNING) {
+            return "경고";
+        }
+        return "이상무";
```
(단순 질문) 실질적으로 여기에 `NORMAL` 케이스는 안들어오는거죠?

changuii replied:
맞아요!

Rok93 commented on src/main/java/components/SavedDataLoader.java:
```diff
@@ -0,0 +1,57 @@
+                .collect(Collectors.toList()));
+    }
+
+
```
불필요한 줄바꿈은 제거해주세요. 😉 

```suggestion

```

Rok93 commented on src/main/java/dto/AttendanceRecordDto.java:
```diff
@@ -0,0 +1,12 @@
+
+public record AttendanceRecordDto(
+        LocalDate attendanceDate,
+        Optional<LocalTime> attendanceTime,
```
Optional을 필드에 사용할 목적으로 만들어진 것이 아니기 때문에 사용하지 않으면 좋을 것 같아요. 😃

[이 글](https://homoefficio.github.io/2019/10/03/Java-Optional-%EB%B0%94%EB%A5%B4%EA%B2%8C-%EC%93%B0%EA%B8%B0/)을 참고해보시면 좋을 것 같습니다.

changuii replied:
dto가 Data Transfer Object인 만큼, 최대한 기능이 없게 만드는데 집중하다 보니 이렇게 작성했습니다.
Optional은 내부에서의 null을 외부로 전달할때만, 사용하는 군요! 😃

Rok93 commented on src/main/java/utils/DayOfWeeks.java:
```diff
@@ -0,0 +1,13 @@
+
+import java.time.DayOfWeek;
+
+public class DayOfWeeks {
```
이런 케이스는 네이밍이 겹치면 오해의 여지가 생길 수도 있기 때문에, 클래스 네이밍을 확실하게 구분해주는게 조금 더 좋다고 생각해요. 

```suggestion
public class DayOfWeekUtils {
```

changuii replied:
저는 Objects, Collections, Arrays와 같이 유틸성 클래스에 붙이는 관용적인 표현이라 이렇게 명명했는데 오히려 많이 사용하지 않는 관습인가 보네요!
반영하겠습니다! :)

Rok93 commented on src/test/java/domain/AttendanceBookTest.java:
```diff
@@ -0,0 +1,144 @@
+import org.junit.jupiter.params.ParameterizedTest;
+import org.junit.jupiter.params.provider.ValueSource;
+
+public class AttendanceBookTest {
```
커밋 이력을 확인했을 때, TDD 사이클에 맞춰 잘 수행해주신 것으로 보여요.  👍💯
테스트 코드 또한 전체적으로 꼼꼼하게 잘 작성해주셨네요. 👍👍👍

Rok93 reviewed (changes requested):
부기 피드백 반영 잘해주셨습니다. 👍💯
지금도 충분히 잘 수행해주셨지만, 
마지막으로 짚고 넘어가면 좋을 부분이보여서 코멘트 남겨두었어요. 
마지막으로 확인 한번 부탁드릴게요. 😉

changuii committed:
refactor: 불필요한 줄바꿈을 제거하라

changuii committed:
refactor(AttendanceHistory): 날짜와 시간에 대한 검증을 묶어서 처리하라

- 실수로 검증하지 않는 상황을 미연에 방지하기 위해 같이 사용되는 validate를 동시에 수행하도록 했다.

changuii committed:
refactor(AttendanceRecordDto): Optional 필드를 제거하라

changuii committed:
refactor(DayOfWeekUtils): DayOfWeeks보다 더 직관적으로 유틸 클래스로 이해할 수 있는, DayOfWeekUtils로 변경하라

changuii committed:
refactor: domain -> Dto 변환의 책임을 dto가 가지도록 하라

changuii committed:
fix(Crews): Crews에 중복 크루가 등록되는 문제를 해결하라

- 추가적으로 방어적 복사도 적용하였다.

changuii committed:
refactor: SavedDataLoad의 패키지를 domain으로 옮겨라

- 데이터를 읽어와 도메인을 만드는 로직을 가지고 있기 때문에 충분히 도메인 로직이라 판단하였다.

changuii committed:
refactor(AttendanceSavedDataLoader): SavedDataLoader보다는 더 직관적인 이름으로 변경하였다.

changuii committed:
refactor: 도메인을 필드로 가지지 않도록 변경하였다.

changuii reviewed (commented):
### 변경점
1. AttendanceHistory에서 날짜와 시간에 대한 검증을 묶어서 처리하도록 변경하여 실수를 방지하도록 하였습니다.
2. Optional 필드를 제거하여, 의도에 맞게 사용하였습니다.
3. DayOfWeeks보다는 더 유틸 클래스에 직관적인 DayOfWeekUtils로 변경하였습니다.
4. DTO 변환의 책임을 DTO가 가지도록 하였습니다.
5. Crews에 크루가 중복 등록되는 문제를 해결하였습니다.
6. SavedDataLoader를 AttendanceSavedDataLoader로 이름을 변경하고 도메인 패키지로 옮겼습니다.

Rok93 reviewed (approved):
부기 피드백 반영 잘해주셨습니다. 👍💯
2단계도 충분히 잘해주신 것 같아서, 출석 미션은 이만 종료하도록하겠습니다. 🎉🎉 
다음 미션도 화이팅입니다. 💪
